src/server/handlers/launchpad.js

```diff
@@ -0,0 +1,236 @@
+  }
+};
+
+const RESPONSE_GITHUB_OTHER = {
+  status: 'error',
+  payload: {
+    code: 'github-error-other',
+    message: 'Something went wrong when looking for snapcraft.yaml'
+  }
+};
+
+const RESPONSE_SNAPCRAFT_YAML_PARSE_FAILED = {
+  status: 'error',
+  payload: {
+    code: 'snapcraft-yaml-parse-failed',
+    message: 'Failed to parse snapcraft.yaml'
+  }
+};
+
+const RESPONSE_SNAPCRAFT_YAML_NO_NAME = {
+  status: 'error',
+  payload: {
+    code: 'snapcraft-yaml-no-name',
+    message: 'snapcraft.yaml has no top-level "name" attribute'
+  }
+};
+
+const RESPONSE_SNAP_NOT_FOUND = {
+  status: 'error',
+  payload: {
+    code: 'snap-not-found',
+    message: 'Cannot find existing snap based on this URL'
+  }
+};
+
+let memcached = null;
+
+const getMemcached = () => {
+  if (memcached === null) {
+    memcached = new Memcached(conf.get('MEMCACHED_HOST').split(','),
+                              { namespace: 'lp:' });
+  }
+  return memcached;
+};
+
+// Test affordance.
+export const setMemcached = (value) => {
+  memcached = value;
+};
+
+const makeSnapName = url => {
+  return createHash('md5').update(url).digest('hex');
+};
+
+const getSnapcraftYaml = (req, res, callback) => {
+  const repositoryUrl = req.body.repository_url;
+  const parsed = parseGitHubUrl(repositoryUrl);
+  if (parsed === null || parsed.owner === null || parsed.name === null) {
+    logger.info(`Cannot parse "${repositoryUrl}"`);
+    return res.status(400).send(RESPONSE_GITHUB_BAD_URL);
+  }
+
+  const uri = `/repos/${parsed.owner}/${parsed.name}/contents/snapcraft.yaml`;
+  const options = {
+    headers: {
+      'Authorization': `token ${req.session.token}`,
+      'Accept': 'application/vnd.github.v3.raw'
+    }
+  };
+  logger.info(`Fetching snapcraft.yaml from ${repositoryUrl}`);
+  return requestGitHub.get(uri, options, (err, response, body) => {
+    if (response.statusCode !== 200) {
+      try {
+        body = JSON.parse(body);
+      } catch (e) {
+        logger.info('Invalid JSON received', err, body);
+        return res.status(500).send(RESPONSE_GITHUB_OTHER);
+      }
+      switch (body.message) {
+        case 'Not Found':
+          // snapcraft.yaml not found
+          return res.status(404).send(RESPONSE_GITHUB_NOT_FOUND);
+        case 'Bad credentials':
+          // Authentication failed
+          return res.status(401).send(RESPONSE_GITHUB_AUTHENTICATION_FAILED);
+        default:
+          // Something else
+          logger.info('GitHub API error:', err, body);
+          return res.status(500).send(RESPONSE_GITHUB_OTHER);
+      }
+    }
+
+    let snapcraftYaml;
+    try {
+      snapcraftYaml = yaml.safeLoad(body);
+    } catch (e) {
+      return res.status(400).send(RESPONSE_SNAPCRAFT_YAML_PARSE_FAILED);
+    }
+    return callback(snapcraftYaml);
+  });
+};
+
+// XXX cjwatson 2016-12-13: We should set an appropriate set of
+// architectures.
+export const newSnap = (req, res) => {
+  getSnapcraftYaml(req, res, snapcraftYaml => {
```

Comment:
It's not a huge deal (especially now in first phase), but we have a bit of duplication. We already call GH API for snapcraft.yaml when verifying user input in https://github.com/canonical-ols/build.snapcraft.io/blob/master/src/common/actions/repository-input.js#L33

Up there we don't parse the response (we just check if the file exists), so for sure we need this raw response to parse it here, but ideally we should only ask the API once.

@cjwatson @katiefenn We are getting to the point where we have some 'loose' pieces without having a bigger picture how all of them would work together in the end.
We have GH and SSO authentication, GH repo input and verification, getting data from LP and creating new snap/build, then listing builds...

It would be worth having some kind of flow chart of the whole user path through the app. It would clarify both user path and all the internal and external (to LP, GH) API calls we need on the way.